File: Sources/FloatingLyricCore/Playback/NowPlayingPoller.swift
import Foundation

public final class NowPlayingPoller: @unchecked Sendable {
    public static let playingIntervalMs = 3_000
    public static let idleIntervalMs = 10_000

    private static let endpoint =
        URL(string: "https://api.spotify.com/v1/me/player/currently-playing")!

    private let auth: SpotifyAuth
    private let http: HTTPClient
    private var loop: Task<Void, Never>?

    public init(auth: SpotifyAuth, http: HTTPClient) {
        self.auth = auth
        self.http = http
    }

    public static func intervalMs(for state: PlaybackState) -> Int {
        if case .playing = state { return playingIntervalMs }
        return idleIntervalMs
    }

    public func start(onState: @escaping @Sendable (PlaybackState) -> Void) {
        stop()
        loop = Task { [weak self] in
            while !Task.isCancelled {
                guard let self else { return }
                let state = await self.pollOnce()
                onState(state)

                var waitMs = Self.intervalMs(for: state)
                if case .failed(.rateLimited(let retryAfter)) = state {
                    waitMs = max(waitMs, Int(retryAfter * 1000))
                }
                try? await Task.sleep(nanoseconds: UInt64(waitMs) * 1_000_000)
            }
        }
    }

    public func stop() {
        loop?.cancel()
        loop = nil
    }

    public func pollOnce() async -> PlaybackState {
        do {
            let response = try await request()
            if response.status == 401 {
                auth.invalidateAccessToken()
                let retry = try await request()
                guard retry.status != 401 else { return .failed(.sessionExpired) }
                return decode(retry)
            }
            return decode(response)
        } catch let error as AppError {
            return .failed(error)
        } catch {
            return .failed(.network(error.localizedDescription))
        }
    }

    private func request() async throws -> HTTPResponse {
        var request = URLRequest(url: Self.endpoint)
        request.setValue("Bearer \(try await auth.accessToken())",
                         forHTTPHeaderField: "Authorization")
        return try await http.send(request)
    }

    private func decode(_ response: HTTPResponse) -> PlaybackState {
        if response.status == 204 || response.body.isEmpty { return .idle }
        if response.status == 429 {
            let retryAfter = TimeInterval(response.headers["Retry-After"] ?? "") ?? 5
            return .failed(.rateLimited(retryAfter: retryAfter))
        }
        guard (200..<300).contains(response.status) else {
            return .failed(.badResponse(status: response.status))
        }
        guard let payload = try? JSONDecoder().decode(CurrentlyPlaying.self, from: response.body),
              let item = payload.item,
              let id = item.id, let name = item.name, let duration = item.duration_ms,
              let artist = item.artists?.first?.name
        else { return .idle }

        let track = TrackIdentity(id: id, title: name, artist: artist,
                                  album: item.album?.name ?? "", durationMs: duration)
        let np = NowPlaying(track: track,
                            progressMs: payload.progress_ms ?? 0,
                            isPlaying: payload.is_playing ?? false,
                            albumArtURL: item.album?.images?.first?.url
                                .flatMap(URL.init(string:)))
        return np.isPlaying ? .playing(np) : .paused(np)
    }

    private struct CurrentlyPlaying: Decodable {
        struct Artist: Decodable { let name: String? }
        struct Image: Decodable { let url: String? }
        struct Album: Decodable { let name: String?; let images: [Image]? }
        struct Item: Decodable {
            let id: String?
            let name: String?
            let duration_ms: Int?
            let artists: [Artist]?
            let album: Album?
        }
        let is_playing: Bool?
        let progress_ms: Int?
        let item: Item?
    }
}

File: windows/src/FloatingLyric.Core/NowPlayingPoller.cs
using System.Text.Json;

namespace FloatingLyric.Core;

/// <summary>
/// The slow loop: asks Spotify what is playing, every few seconds. Between
/// polls, <see cref="PlayheadClock"/> fills in the gaps.
/// </summary>
public sealed class NowPlayingPoller(SpotifyAuth auth, IHttpClient http)
{
    public const int PlayingIntervalMs = 3_000;
    public const int IdleIntervalMs = 10_000;

    private const string Endpoint = "https://api.spotify.com/v1/me/player/currently-playing";

    private CancellationTokenSource? _loop;

    public static int IntervalMs(PlaybackState state) =>
        state is PlaybackState.Playing ? PlayingIntervalMs : IdleIntervalMs;

    public void Start(Action<PlaybackState> onState)
    {
        Stop();
        var cts = new CancellationTokenSource();
        _loop = cts;

        _ = Task.Run(async () =>
        {
            while (!cts.IsCancellationRequested)
            {
                var state = await PollOnceAsync();
                if (cts.IsCancellationRequested) return;
                onState(state);

                var waitMs = IntervalMs(state);
                if (state is PlaybackState.Failed { Error: AppError.RateLimited limited })
                    waitMs = Math.Max(waitMs, (int)(limited.RetryAfterSeconds * 1000));

                try { await Task.Delay(waitMs, cts.Token); }
                catch (TaskCanceledException) { return; }
            }
        }, cts.Token);
    }

    public void Stop()
    {
        _loop?.Cancel();
        _loop = null;
    }

    public async Task<PlaybackState> PollOnceAsync()
    {
        try
        {
            var response = await RequestAsync();
            if (response.Status == 401)
            {
                auth.InvalidateAccessToken();
                var retry = await RequestAsync();
                if (retry.Status == 401)
                    return new PlaybackState.Failed(new AppError.SessionExpired());
                return Decode(retry);
            }
            return Decode(response);
        }
        catch (AppErrorException e)
        {
            return new PlaybackState.Failed(e.Error);
        }
        catch (Exception e)
        {
            return new PlaybackState.Failed(new AppError.Network(e.Message));
        }
    }

    private async Task<HttpResponse> RequestAsync()
    {
        using var request = new HttpRequestMessage(HttpMethod.Get, Endpoint);
        request.Headers.Add("Authorization", $"Bearer {await auth.AccessTokenAsync()}");
        return await http.SendAsync(request);
    }

    private static PlaybackState Decode(HttpResponse response)
    {
        // Status first, body second. Spotify sends 429 with an empty body, so
        // checking emptiness first reads rate limiting as "nothing playing" —
        // throwing away Retry-After and polling straight back into the limit.
        if (response.Status == 429)
        {
            var retryAfter = response.Headers.TryGetValue("Retry-After", out var value)
                && double.TryParse(value, out var seconds) ? seconds : 5;
            return new PlaybackState.Failed(new AppError.RateLimited(retryAfter));
        }
        if (response.Status is < 200 or >= 300)
            return new PlaybackState.Failed(new AppError.BadResponse(response.Status));
        if (response.Status == 204 || response.Body.Length == 0) return new PlaybackState.Idle();

        try
        {
            using var json = JsonDocument.Parse(response.Body);
            var root = json.RootElement;
            if (!root.TryGetProperty("item", out var item) ||
                item.ValueKind != JsonValueKind.Object) return new PlaybackState.Idle();

            var id = item.GetPropertyOrNull("id")?.GetString();
            var name = item.GetPropertyOrNull("name")?.GetString();
            var duration = item.GetPropertyOrNull("duration_ms")?.GetInt32();
            var artist = item.GetPropertyOrNull("artists") is { ValueKind: JsonValueKind.Array } artists
                && artists.GetArrayLength() > 0
                ? artists[0].GetPropertyOrNull("name")?.GetString()
                : null;

            if (id is null || name is null || duration is null || artist is null)
                return new PlaybackState.Idle();

            var album = item.GetPropertyOrNull("album");
            var track = new TrackIdentity(id, name, artist,
                album?.GetPropertyOrNull("name")?.GetString() ?? "", duration.Value);

            var np = new NowPlaying(
                track,
                root.GetPropertyOrNull("progress_ms")?.GetInt32() ?? 0,
                root.GetPropertyOrNull("is_playing")?.GetBoolean() ?? false,
                album?.GetPropertyOrNull("images") is { ValueKind: JsonValueKind.Array } images
                    && images.GetArrayLength() > 0
                    ? images[0].GetPropertyOrNull("url")?.GetString()
                    : null);

            return np.IsPlaying ? new PlaybackState.Playing(np) : new PlaybackState.Paused(np);
        }
        catch (JsonException)
        {
            return new PlaybackState.Idle();
        }
    }
}

internal static class JsonExtensions
{
    public static JsonElement? GetPropertyOrNull(this JsonElement element, string name) =>
        element.ValueKind == JsonValueKind.Object &&
        element.TryGetProperty(name, out var value) &&
        value.ValueKind != JsonValueKind.Null
            ? value : null;
}
